core/math/dynamic_bvh.cpp

```diff
@@ -0,0 +1,400 @@
+			--end;
+		}
+
+		if (begin == end) {
+			break;
+		}
+
+		// swap out of place nodes
+		--end;
+		Node *temp = leaves[begin];
+		leaves[begin] = leaves[end];
+		leaves[end] = temp;
+		++begin;
+	}
+
+	return begin;
+}
+
+DynamicBVH::Volume DynamicBVH::_bounds(Node **leaves, int p_count) {
+	Volume volume = leaves[0]->volume;
+	for (int i = 1, ni = p_count; i < ni; ++i) {
+		volume = volume.merge(leaves[i]->volume);
+	}
+	return (volume);
+}
+
+void DynamicBVH::_bottom_up(Node **leaves, int p_count) {
+	while (p_count > 1) {
+		real_t minsize = Math_INF;
+		int minidx[2] = { -1, -1 };
+		for (int i = 0; i < p_count; ++i) {
+			for (int j = i + 1; j < p_count; ++j) {
+				const real_t sz = leaves[i]->volume.merge(leaves[j]->volume).get_size();
+				if (sz < minsize) {
+					minsize = sz;
+					minidx[0] = i;
+					minidx[1] = j;
+				}
+			}
+		}
+		Node *n[] = { leaves[minidx[0]], leaves[minidx[1]] };
+		Node *p = _create_node_with_volume(nullptr, n[0]->volume.merge(n[1]->volume), nullptr);
+		p->childs[0] = n[0];
+		p->childs[1] = n[1];
+		n[0]->parent = p;
+		n[1]->parent = p;
+		leaves[minidx[0]] = p;
+		leaves[minidx[1]] = leaves[p_count - 1];
+		--p_count;
+	}
+}
+
+DynamicBVH::Node *DynamicBVH::_top_down(Node **leaves, int p_count, int p_bu_treshold) {
+	static const Vector3 axis[] = { Vector3(1, 0, 0), Vector3(0, 1, 0), Vector3(0, 0, 1) };
+
+	ERR_FAIL_COND_V(p_bu_treshold <= 1, nullptr);
+	if (p_count > 1) {
+		if (p_count > p_bu_treshold) {
+			const Volume vol = _bounds(leaves, p_count);
+			const Vector3 org = vol.get_center();
+			int partition;
+			int bestaxis = -1;
+			int bestmidp = p_count;
+			int splitcount[3][2] = { { 0, 0 }, { 0, 0 }, { 0, 0 } };
+			int i;
+			for (i = 0; i < p_count; ++i) {
+				const Vector3 x = leaves[i]->volume.get_center() - org;
+				for (int j = 0; j < 3; ++j) {
+					++splitcount[j][x.dot(axis[j]) > 0 ? 1 : 0];
+				}
+			}
+			for (i = 0; i < 3; ++i) {
+				if ((splitcount[i][0] > 0) && (splitcount[i][1] > 0)) {
+					const int midp = (int)Math::abs(real_t(splitcount[i][0] - splitcount[i][1]));
+					if (midp < bestmidp) {
+						bestaxis = i;
+						bestmidp = midp;
+					}
+				}
+			}
+			if (bestaxis >= 0) {
+				partition = _split(leaves, p_count, org, axis[bestaxis]);
+				ERR_FAIL_COND_V(partition == 0 || partition == p_count, nullptr);
+			} else {
+				partition = p_count / 2 + 1;
+			}
+
+			Node *node = _create_node_with_volume(nullptr, vol, nullptr);
+			node->childs[0] = _top_down(&leaves[0], partition, p_bu_treshold);
+			node->childs[1] = _top_down(&leaves[partition], p_count - partition, p_bu_treshold);
+			node->childs[0]->parent = node;
+			node->childs[1]->parent = node;
+			return (node);
+		} else {
+			_bottom_up(leaves, p_count);
+			return (leaves[0]);
+		}
+	}
+	return (leaves[0]);
+}
+
+DynamicBVH::Node *DynamicBVH::_node_sort(Node *n, Node *&r) {
+	Node *p = n->parent;
+	ERR_FAIL_COND_V(!n->is_internal(), nullptr);
+	if (p > n) {
```

Comment:
Is this meant to compare the pointers or are we missing a comparison operator?

Reply:
Good catch, it appears to be like this on Bullet too, it does not make much sense to me, I wonder why is this done this way..

Reply:
bullet also allocates on heap, so I should assume it has some sort of meaning, but i really really cant figure it out.

Reply:
yup, I would expect the comparison to have some sort of geometric meaning...

Reply:
My wild guess is that it may be to avoid a redundant rotation or something like this, but the code is too cryptic. Maybe @lawnjelly knows?

Reply:
Maybe it's a variation of `!=`. It doesn't make much sense.

Reply:
Could it be related to cache efficiency? With this sorting the nodes closer to the root of the tree will end up closer in memory, and depending on the allocator they could even be in contiguous memory.

Reply:
It looks like this is just sorting nodes by memory address for cache optimization. It swaps parent and child completely, including their volumes so I don't think it has a geometric reason.

Reply:
Ah, I had missed the latest comment. I agree with this interpretation :)